libraries/node-core-library/src/Executable.ts

```diff
@@ -0,0 +1,419 @@
+
+      if (Executable._canExecute(resolvedPath, context)) {
+        return resolvedPath;
+      }
+
+      // Try the default file extensions
+      for (const shellExtension of context.windowsExecutableExtensions) {
+        const resolvedNameWithExtension: string = resolvedPath + shellExtension;
+
+        if (Executable._canExecute(resolvedNameWithExtension, context)) {
+          return resolvedNameWithExtension;
+        }
+      }
+    }
+
+    return undefined;
+  }
+
+  /**
+   * This is used when searching the shell PATH for an executable, to determine
+   * whether a match should be skipped or not.  If it returns true, this does not
+   * guarantee that the file can be successfully executed.
+   */
+  private static _canExecute(filePath: string, context: IExecutableContext): boolean {
+    if (!FileSystem.exists(filePath)) {
+      return false;
+    }
+
+    const fileExtension: string = path.extname(filePath);
+
+    if (os.platform() === 'win32') {
+      // Does the file have an executable file extension?
+      let matchFound: boolean = false;
+      for (const executableExtension of context.windowsExecutableExtensions) {
+        if (fileExtension.localeCompare(executableExtension) === 0) {
+          matchFound = true;
+          break;
+        }
+      }
+
+      if (fileExtension.localeCompare('.js') === 0) {
+        // The .js extension is special and always allowed since this is NodeJS.
+        // However, that file extension must be specified explitily unless it was added to PATHEXT
+        // (which is not usually the case).
+        matchFound = true;
+      }
+
+      if (!matchFound) {
+        return false;
+      }
+    } else {
+      // For Unix, check whether any of the POSIX execute bits are set, but don't
+      // require this for JavaSCript scripts (since this is NodeJS)
+      if (fileExtension.localeCompare('.js') !== 0) {
+        try {
+          // tslint:disable-next-line:no-bitwise
+          if ((FileSystem.getPosixModeBits(filePath) & PosixModeBits.AllExecute) === 0) {
+            return false; // not executable
+          }
+        } catch (error) {
+          // If we have trouble accessing the file, ignore the error and consider it "not executable"
+          // since that's what a shell would do
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns the list of folders where we will search for an executable,
+   * based on the PATH environment variable.
+   */
+  private static _getSearchFolders(context: IExecutableContext): string[] {
+
+    const pathList: string = context.environment.PATH || '';
+
+    const folders: string[] = [];
+
+    // Avoid processing duplicates
+    const seenPaths: Set<string> = new Set<string>();
+
+    if (os.platform() === 'win32') {
+      // On Window, the current directory is always tried first
+      folders.push(context.currentWorkingDirectory);
+      seenPaths.add(context.currentWorkingDirectory);
```

Comment:
Do we want to allow this? My preference would be to lean towards the POSIX ./ requirement.